src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1391 @@
+                }
+                self.available_pregs[OperandPos::Late].remove(preg);
+            }
+            _ => {
+                if op.as_fixed_nonallocatable().is_none()
+                    && (!early_avail_pregs.contains(preg) || !late_avail_pregs.contains(preg))
+                {
+                    return Err(RegAllocError::TooManyLiveRegs);
+                }
+                self.available_pregs[OperandPos::Early].remove(preg);
+                self.available_pregs[OperandPos::Late].remove(preg);
+            }
+        }
+        Ok(())
+    }
+
+    fn allocd_within_constraint(&self, op: Operand) -> bool {
+        let alloc = self.vreg_allocs[op.vreg().vreg()];
+        match op.constraint() {
+            OperandConstraint::Any => {
+                if let Some(preg) = alloc.as_reg() {
+                    if !self.available_pregs[op.pos()].contains(preg) {
+                        // If a register isn't in the available pregs list, then
+                        // there are two cases: either it's reserved for a
+                        // fixed register constraint or a vreg allocated in the instruction
+                        // is already assigned to it.
+                        //
+                        // For example:
+                        // 1. use v0, use v0, use v0
+                        //
+                        // Say p0 is assigned to v0 during the processing of the first operand.
+                        // When the second v0 operand is being processed, v0 will still be in
+                        // v0, so it is still allocated within constraints.
+                        trace!("The vreg in {preg}: {}", self.vreg_in_preg[preg.index()]);
+                        self.vreg_in_preg[preg.index()] == op.vreg()
+                    } else {
+                        true
+                    }
+                } else {
+                    !alloc.is_none()
+                }
+            }
+            OperandConstraint::Reg => {
+                if self.is_stack(alloc) {
+                    return false;
+                }
+                if let Some(preg) = alloc.as_reg() {
+                    if !self.available_pregs[op.pos()].contains(preg) {
+                        trace!("The vreg in {preg}: {}", self.vreg_in_preg[preg.index()]);
+                        self.vreg_in_preg[preg.index()] == op.vreg()
+                    } else {
+                        true
+                    }
+                } else {
+                    false
+                }
+            }
+            // It is possible for an operand to have a fixed register constraint to
+            // a clobber.
+            OperandConstraint::FixedReg(preg) => alloc.is_reg() && alloc.as_reg().unwrap() == preg,
+            OperandConstraint::Reuse(_) => {
+                unreachable!()
+            }
+        }
+    }
+
+    fn evict_vreg_in_preg_before_inst(&mut self, inst: Inst, preg: PReg) {
```

Comment:
It looks like this function and `evict_vreg_in_preg` below are the same except for the `InstPosition` on the move, is that right? Can we pass it in as a parameter and deduplicate the two functions?